---
title: "1. Create reference directory for mapping"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{1. Create reference directory for mapping}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc} 
---
  
`scrattch.mapping` offers functions to convert completed single cell/nucleus RNA-seq analysis (e.g., a counts matrix + cell type assignments) into a standard reference taxonomy compatible with various mapping techniques, as well as functions to perform such mapping analyses.  Several functions are also provided for conversion of output results into data and folder formats compatible with Allen Institute R shiny tools.  
  
The `scrattch.mapping` package is one component of the scrattch suite of packages for `S`ingle `C`ell `R`NA-seq `A`nalysis for `T`ranscriptomic `T`ype `CH`aracterization from the Allen Institute.  
  
For this vignette, we use the dataset published in [Tasic, et al. (2016) Nature Neuroscience](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4985242/), which is available in the [tasic2016data package](https://github.com/AllenInstitute/tasic2016data):  
  
```{r install data package if needed}
if(!"tasic2016data" %in% rownames(installed.packages())) {
  devtools::install_github("AllenInstitute/tasic2016data")
}
suppressPackageStartupMessages({
  library(tasic2016data)
})
```
  
  
This vignette also creates taxonomy standards published in [Miller, et al. (2020) eLife](https://elifesciences.org/articles/59928) using the [CCN R package](https://github.com/AllenInstitute/CCN).
  
```{r install CCN if needed}
if(!"CCN" %in% rownames(installed.packages())) {
  ## Imported packages
  packages <- c("remotes","jsonlite","data.table","ggplot2","dendextend","dplyr","BiocManager")
  packages <- packages[!(packages %in% installed.packages()[,"Package"])]
  if(length(packages)>0) install.packages(new.packages, repos='http://cran.us.r-project.org')
  BiocManager::install("rols", update=FALSE)

  ## CCN install
  remotes::install_github("AllenInstitute/CCN", build_vignettes = TRUE)
}
suppressPackageStartupMessages({
  library(CCN)
})
```
  
  
### Workspace set-up  
  
First, let's create a directory in the current working directory to store the reference taxonomy files.  
  
```{r create reference directory}
# Replace link below with link to directory
taxonomyDir <- paste0(system.file(package = 'scrattch.mapping'),"/reference/") 
dir.create(taxonomyDir, showWarnings = FALSE)
print(taxonomyDir)
```

  
Load libraries and set options.  
  
```{r load libraries, warning=FALSE}
suppressPackageStartupMessages({
  library("scrattch.mapping")
})
options(stringsAsFactors = FALSE)
options(future.globals.maxSize = 4000 * 1024^2)  # Can adjust this value if needed, depending on number of cells
options(future.rng.onMisuse="ignore")
```
  
  
Read in the reference data (in this case we will use the Tasic 2016 data, which includes ~1800 cells from mouse primary visual cortex), exclude unclustered cells, and subsample to 25 cells per cluster (for the purpose of the Vignette--typically you'd retain more than this).  Note that there are several places where subsampling can occur.  Either here or in the buildTaxonomy step will lead to a smaller data set, whereas in the addDendrogramMarkers step will only impact marker genes that end up in the tree nodes.  
  
```{r load tasic data}
annotations <- tasic_2016_anno
counts      <- tasic_2016_counts  # uncomment if using CPM below
annotations <- annotations[match(colnames(counts),annotations$sample_name),]  # Put them in the correct order
rownames(annotations) <- annotations$sample_name  

# Subsampling to speed up Vignette and removing unclassified cells 
kp          <- (annotations$broad_type!="Unclassified")&(subsampleCells(annotations$primary_type_label,25,seed = 1))
counts      <- counts[,kp]
annotations <- annotations[kp,]

# Remove all genes with no expression to speed up Vignette
counts      <- counts[rowSums(counts)>0,]
```
  
  
Now we are going to rename some of the annotations to match current Allen Institute terminology.  This is not strictly necessary, but is useful.  
  
```{r add level annotations}
# Rename the primary cell type to "cluster"
clusters            <- annotations$primary_type_label[match(1:49,annotations$primary_type_id)]
annotations$cluster <- factor(annotations$primary_type_label, levels=clusters)

# Retain cluster colors
cluster_colors <- setNames(annotations$primary_type_color[match(1:49,annotations$primary_type_id)], clusters)

# Create a class label that groups all non-neurons together
annotations$class <- annotations$broad_type
annotations$class[!is.element(annotations$class,c("GABA-ergic Neuron","Glutamatergic Neuron"))] = "Non-neuronal"

# Create a subclass label (e.g., intermediate resolution between class and cell type)
subclass <- c(rep("CGE",9),rep("MGE",13),"CGE",rep("Exc IT",10),rep("Exc non-IT",9),rep("NN",7))
subclass <- setNames(subclass,annotations$cluster[match(1:49,annotations$cluster_id)])
subclass <- subclass[annotations$cluster]
subclass[annotations$class=="Non-neuronal"] <- annotations$broad_type[annotations$class=="Non-neuronal"]
annotations$subclass <- gsub("Oligodendrocyte Precursor Cell","OPC",subclass)
head(data.frame(annotations[,c("cluster","subclass","class")]))
```
  
  
Finally, prior to generating the taxonomy, we need to define a set of variable genes and build a 2-dimensional representation of the data.  To do this we use binary genes and the `umap` R library, respectively, although other options could also be used (for example, the Seurat R package often provides comparable or better looking results).  
  
```{r create UMAP representation, fig.height=8, fig.width=10}
## Get top 1000 genes by a beta (binary) score
binary.genes <- top_binary_genes(counts,annotations$cluster,1000)

## Calculate a UMAP based on these binary genes.
npcs <- 30
pcs  <- prcomp(logCPM(counts)[binary.genes,], scale = TRUE)$rotation
umap.coords <- umap(pcs[,1:npcs])$layout

## View the output as a sanity check
plot(umap.coords[,1],umap.coords[,2],col=annotations$primary_type_color, pch=19, cex=0.5,
     main="",xlab="UMAP 1", ylab="UMAP 2")
```
  
  
## Generate the reference directory
  
For the first step we convert the counts and annotations variables into a Seurat object of a particular format.  This is then used as input into functions for generating files in the reference directory and for creating a dendrogram.
 
Next, create the reference folder with all associated files, and a preliminary dendrogram.  The dendrogram will get updated in later functions.
  
```{r build reference folder} 
buildTaxonomy(counts         = counts,
              meta.data      = annotations,
              feature.set    = binary.genes,
              umap.coords    = umap.coords,
              taxonomyDir    = taxonomyDir,
              cluster_colors = cluster_colors,  # Colors are generated if not provided
              metadata_names = NULL,     # If you want to save only selected metadata columns
              subsample      = 25,      # This is the number of cells per cluster (in this case redundant with above)
              reorder.dendrogram = TRUE) # This will match the factor order of the clusters to the extent possible
```

We repackage this information to a compact annData format compatible with both R and python.  **This annData object is the input file for all mapping algorithms in this package, and a to-be-created parallel python package.**.  
```{r create annData object}
AIT.anndata <- loadTaxonomy(taxonomyDir, hGenes=binary.genes)
```
  
Now we annotate the dendrogram with marker genes.  This is only required for using tree mapping, and is critical if this reference is to be used for processing Patch-seq data from the Allen Institute.  For other mapping strategies, this is not needed.  Note that we can either pull data and metadata from variables or directly from the reference folder.  This example pulls from the reference folder.  
  
```{r add dendrogram markers}
dend <- addDendrogramMarkers(AIT.anndata, 
                             subsample= 25) # To speed up calculations a bit, but again redundant with above.
```

Next, we collect some files need for applying patchSeqQC to patch-seq data. This is not needed for mapping, but is an important QC step when comparing  

```{r patchseqQC set up}
AIT.anndata = buildPatchseqTaxonomy(AIT.anndata,
                       subsample = 100,  # Default of 100 is reasonable
                       subclass.column = "subclass_label",  # default
                       class.column = "class_label",  # default
                       off.target.types = "Non-neuronal",  # default is various iterations of non-neuronal
                       num.markers = 50,     # Default of 50 is probably fine
                       taxonomyDir = taxonomyDir)
```
  
Output session info.  
  
```{r session info}
sessionInfo()
```

