<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>2. Map patch-seq data and output directory</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">2. Map patch-seq data and output directory</h1>



<p><code>scrattch.mapping</code> offers functions to convert completed single cell/nucleus RNA-seq analysis (e.g., a counts matrix + cell type assignments) into a standard reference taxonomy compatible with various mapping techniques, as well as functions to perform such mapping analyses. Several functions are also provided for conversion of output results into data and folder formats compatible with Allen Institute R shiny tools.</p>
<p>The <code>scrattch.mapping</code> package is one component of the scrattch suite of packages for <code>S</code>ingle <code>C</code>ell <code>R</code>NA-seq <code>A</code>nalysis for <code>T</code>ranscriptomic <code>T</code>ype <code>CH</code>aracterization from the Allen Institute.</p>
<p>This vignette goes through how to map a small data set against a reference taxonomy. Here we use a subset of tasic2016data as an example but the intention is for mapping of patch-seq data. <strong>The vignette “1. Create reference directory for mapping” must be run prior to running this vignette, as the output reference taxonomy files are required for mapping.</strong></p>
<div id="workspace-set-up" class="section level3">
<h3>Workspace set-up</h3>
<p>First, let’s create a directory in the current working directory to store the reference taxonomy files.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># REPLACE THIS LINE WITH the reference directory folder from &quot;Map patch-seq data and output directory&quot;</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>refFolder <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">system.file</span>(<span class="at">package =</span> <span class="st">&#39;scrattch.mapping&#39;</span>),<span class="st">&quot;/reference/&quot;</span>)  </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Replase this line with wherever you&#39;d like your mapping output</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>mappingFolder <span class="ot">&lt;-</span> <span class="fu">paste0</span>(refFolder,<span class="st">&quot;/mapping/&quot;</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(mappingFolder, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(mappingFolder)</span></code></pre></div>
<pre><code>## [1] &quot;C:/Users/jeremym/AppData/Local/Temp/RtmpMbYaYx/Rinst66ec45977828/scrattch.mapping/reference//mapping/&quot;</code></pre>
<p>Load libraries and set options.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">&quot;scrattch.mapping&quot;</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">&quot;tasic2016data&quot;</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">4000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)  <span class="co"># Can adjust this value if needed, depending on number of cells</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.rng.onMisuse=</span><span class="st">&quot;ignore&quot;</span>)</span></code></pre></div>
<p>Read in the reference taxonomy. This can be any reference, but in this case we load the one based from tasic2016data generated in the “1. Create reference directory for mapping” vignette.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>AIT.anndata <span class="ot">&lt;-</span> <span class="fu">read_h5ad</span>(<span class="fu">file.path</span>(refFolder,<span class="st">&quot;reference.h5ad&quot;</span>))</span></code></pre></div>
<p>Read in the data for mapping. In this case we are going to take 5 cells per neuronal cluster from tasic2016data, which is a subset of the data used in “1. Create reference directory for mapping”.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in data</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> tasic_2016_anno</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>counts      <span class="ot">&lt;-</span> tasic_2016_counts  <span class="co"># uncomment if using CPM below</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> annotations[<span class="fu">match</span>(<span class="fu">colnames</span>(counts),annotations<span class="sc">$</span>sample_name),]  <span class="co"># Put them in the correct order</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(annotations) <span class="ot">&lt;-</span> annotations<span class="sc">$</span>sample_name  </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>kp          <span class="ot">&lt;-</span> annotations<span class="sc">$</span>broad_type<span class="sc">!=</span><span class="st">&quot;Unclassified&quot;</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>counts      <span class="ot">&lt;-</span> counts[,kp]</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> annotations[kp,]</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Variable renaming</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>clusters            <span class="ot">&lt;-</span> annotations<span class="sc">$</span>primary_type_label[<span class="fu">match</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">49</span>,annotations<span class="sc">$</span>primary_type_id)]</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>annotations<span class="sc">$</span>cluster <span class="ot">&lt;-</span> <span class="fu">factor</span>(annotations<span class="sc">$</span>primary_type_label, <span class="at">levels=</span>clusters)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>annotations<span class="sc">$</span>class   <span class="ot">&lt;-</span> annotations<span class="sc">$</span>broad_type</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>annotations<span class="sc">$</span>class[<span class="sc">!</span><span class="fu">is.element</span>(annotations<span class="sc">$</span>class,<span class="fu">c</span>(<span class="st">&quot;GABA-ergic Neuron&quot;</span>,<span class="st">&quot;Glutamatergic Neuron&quot;</span>))] <span class="ot">=</span> <span class="st">&quot;Non-neuronal&quot;</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsample to get final results</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>query.kp       <span class="ot">&lt;-</span> <span class="fu">subsampleCells</span>(annotations<span class="sc">$</span>cluster,<span class="at">subSamp=</span><span class="dv">5</span>,<span class="at">seed=</span><span class="dv">10</span>)<span class="sc">&amp;</span>(annotations<span class="sc">$</span>class<span class="sc">!=</span><span class="st">&quot;Non-neuronal&quot;</span>)</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>query.counts   <span class="ot">&lt;-</span> counts[,query.kp]</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>query.metadata <span class="ot">&lt;-</span> annotations[query.kp,]</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>query.logCPM   <span class="ot">&lt;-</span> <span class="fu">logCPM</span>(query.counts)</span></code></pre></div>
<p>The query and reference data sets are now loaded and ready to go!</p>
</div>
<div id="map-query-data" class="section level2">
<h2>Map query data</h2>
<p>The first step for mapping to run the one step “taxonomy_mapping” function to apply Seurat mapping, tree mapping, and correlation mapping. <strong>This step that would be applied for ANY query data.</strong></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>query.mapping <span class="ot">&lt;-</span> <span class="fu">taxonomy_mapping</span>(<span class="at">AIT.anndata=</span> AIT.anndata,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">query.data =</span> query.logCPM, </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">corr.map   =</span> <span class="cn">TRUE</span>, <span class="co"># Flags for which mapping algorithms to run</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">tree.map   =</span> <span class="cn">TRUE</span>, </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">seurat.map =</span> <span class="cn">TRUE</span>, </span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">label.cols =</span> <span class="fu">c</span>(<span class="st">&quot;cluster_label&quot;</span>,<span class="st">&quot;subclass_label&quot;</span>, <span class="st">&quot;class_label&quot;</span>) <span class="co"># Columns to map against</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## [1] &quot;============================== Mapping =======================&quot;
## [1] &quot;Thu Mar 30 15:02:16 2023&quot;
## [1] &quot;Correlation-based mapping&quot;
## [1] &quot;Tree-based mapping&quot;
## [1] &quot;Seurat-based mapping&quot;</code></pre>
<p>The second step is apply the function “buildMappingDirectory” which performs a variety of functions useful for patch-seq analysis:<br />
1. Merge the mapping results into the metadata file.<br />
2. Apply <a href="https://github.com/PavlidisLab/patchSeqQC/">patchseqQC</a> to generate some patch-seq-specific QC metrics for the mapping.<br />
3. Output the results to a directory in a format compatible with Allen Institute visualization tools.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">buildMappingDirectory</span>(<span class="at">AIT.anndata    =</span> AIT.anndata, </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">mappingFolder  =</span> mappingFolder,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">query.data     =</span> query.counts,  <span class="co"># Don&#39;t need log-normalized data here</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">query.metadata =</span> query.metadata,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">query.mapping  =</span> query.mapping,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">doPatchseqQC   =</span> <span class="cn">TRUE</span>  <span class="co"># Set to FALSE if not needed or if writePatchseqQCmarkers was not applied in reference generation</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Joining with `by = join_by(cluster_label)`</code></pre>
<pre><code>## [1] &quot;Colors updated for: primary_type&quot;
## [1] &quot;Colors updated for: cluster_Corr&quot;
## [1] &quot;Colors updated for: cluster_Seurat&quot;
## [1] &quot;Colors updated for: cluster_Tree&quot;
## [1] &quot;Colors updated for: broad_type&quot;
## [1] &quot;Colors updated for: class&quot;
## [1] &quot;Colors updated for: class_Corr&quot;
## [1] &quot;Colors updated for: class_Seurat&quot;
## [1] &quot;Colors updated for: class_Tree&quot;</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir</span>(mappingFolder)</span></code></pre></div>
<pre><code>## [1] &quot;anno.feather&quot;      &quot;data.feather&quot;      &quot;dend.RData&quot;       
## [4] &quot;desc.feather&quot;      &quot;memb.feather&quot;      &quot;tsne.feather&quot;     
## [7] &quot;tsne_desc.feather&quot;</code></pre>
<p>Output session info.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.2.2 (2022-10-31 ucrt)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 10 x64 (build 19042)
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=C                          
## [2] LC_CTYPE=English_United States.utf8   
## [3] LC_MONETARY=English_United States.utf8
## [4] LC_NUMERIC=C                          
## [5] LC_TIME=English_United States.utf8    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] rols_2.20.1           jsonlite_1.8.4        data.table_1.14.8    
##  [4] randomForest_4.7-1.1  scrattch.mapping_0.1  arrow_11.0.0.2       
##  [7] umap_0.2.10.0         cowplot_1.1.1         SeuratObject_4.1.3   
## [10] Seurat_4.3.0          patchSeqQC_0.1.0      patchseqtools_0.0.1  
## [13] mfishtools_0.0.2      scrattch.vis_0.0.210  purrr_1.0.1          
## [16] ggbeeswarm_0.7.1      ggplot2_3.4.1         scrattch.bigcat_0.0.5
## [19] scrattch.hicat_1.0.0  scrattch.io_0.1.0     rhdf5_2.36.0         
## [22] pvclust_2.2-0         foreach_1.5.2         MatrixGenerics_1.4.3 
## [25] matrixStats_0.63.0    Matrix_1.5-3          tibble_3.1.8         
## [28] viridis_0.6.2         viridisLite_0.4.1     dendextend_1.16.0    
## [31] dplyr_1.1.0           anndata_0.7.5.5       feather_0.3.5        
## [34] CCN_0.1.0             tasic2016data_1.0    
## 
## loaded via a namespace (and not attached):
##   [1] rappdirs_0.3.3         scattermore_0.8        tidyr_1.3.0           
##   [4] bit64_4.0.5            knitr_1.42             irlba_2.3.5.1         
##   [7] rpart_4.1.19           KEGGREST_1.32.0        RCurl_1.98-1.10       
##  [10] doParallel_1.0.17      generics_0.1.3         BiocGenerics_0.38.0   
##  [13] preprocessCore_1.54.0  RSQLite_2.3.0          RANN_2.6.1            
##  [16] future_1.31.0          bit_4.0.5              tzdb_0.3.0            
##  [19] spatstat.data_3.0-0    httpuv_1.6.9           assertthat_0.2.1      
##  [22] xfun_0.37              hms_1.1.2              jquerylib_0.1.4       
##  [25] evaluate_0.20          promises_1.2.0.1       fansi_1.0.4           
##  [28] progress_1.2.2         igraph_1.4.1           DBI_1.1.3             
##  [31] htmlwidgets_1.6.1      spatstat.geom_3.0-6    stats4_4.2.2          
##  [34] ellipsis_0.3.2         RSpectra_0.16-1        backports_1.4.1       
##  [37] RcppParallel_5.1.7     deldir_1.0-6           vctrs_0.5.2           
##  [40] Biobase_2.52.0         here_1.0.1             ROCR_1.0-11           
##  [43] abind_1.4-5            cachem_1.0.6           withr_2.5.0           
##  [46] progressr_0.13.0       checkmate_2.1.0        sctransform_0.3.5     
##  [49] prettyunits_1.1.1      goftest_1.2-3          cluster_2.1.4         
##  [52] lazyeval_0.2.2         crayon_1.5.2           spatstat.explore_3.0-6
##  [55] pkgconfig_2.0.3        labeling_0.4.2         GenomeInfoDb_1.28.4   
##  [58] nlme_3.1-162           vipor_0.4.5            nnet_7.3-18           
##  [61] rlang_1.0.6            globals_0.16.2         lifecycle_1.0.3       
##  [64] miniUI_0.1.1.1         rprojroot_2.0.3        polyclip_1.10-4       
##  [67] lmtest_0.9-40          Rhdf5lib_1.14.2        zoo_1.8-11            
##  [70] base64enc_0.1-3        beeswarm_0.4.0         ggridges_0.5.4        
##  [73] png_0.1-8              bitops_1.0-7           KernSmooth_2.23-20    
##  [76] rhdf5filters_1.4.0     Biostrings_2.60.2      blob_1.2.3            
##  [79] stringr_1.5.0          parallelly_1.34.0      spatstat.random_3.1-3 
##  [82] jpeg_0.1-10            S4Vectors_0.30.0       scales_1.2.1          
##  [85] memoise_2.0.1          magrittr_2.0.3         plyr_1.8.8            
##  [88] ica_1.0-3              zlibbioc_1.38.0        compiler_4.2.2        
##  [91] RColorBrewer_1.1-3     fitdistrplus_1.1-8     cli_3.6.0             
##  [94] XVector_0.32.0         listenv_0.9.0          patchwork_1.1.2       
##  [97] pbapply_1.7-0          htmlTable_2.4.1        Formula_1.2-4         
## [100] MASS_7.3-58.2          WGCNA_1.72-1           tidyselect_1.2.0      
## [103] stringi_1.7.12         highr_0.10             yaml_2.3.7            
## [106] askpass_1.1            latticeExtra_0.6-30    ggrepel_0.9.3         
## [109] grid_4.2.2             sass_0.4.5             tools_4.2.2           
## [112] future.apply_1.10.0    parallel_4.2.2         rstudioapi_0.14       
## [115] foreign_0.8-84         gridExtra_2.3          farver_2.1.1          
## [118] Rtsne_0.16             digest_0.6.31          shiny_1.7.4           
## [121] Rcpp_1.0.10            later_1.3.0            RcppAnnoy_0.0.20      
## [124] httr_1.4.5             AnnotationDbi_1.54.1   colorspace_2.1-0      
## [127] tensor_1.5             reticulate_1.28        IRanges_2.26.0        
## [130] splines_4.2.2          uwot_0.1.14            spatstat.utils_3.0-1  
## [133] sp_1.6-0               plotly_4.10.1          xtable_1.8-4          
## [136] dynamicTreeCut_1.63-1  R6_2.5.1               Hmisc_4.8-0           
## [139] pillar_1.8.1           htmltools_0.5.4        mime_0.12             
## [142] glue_1.6.2             fastmap_1.1.0          codetools_0.2-19      
## [145] utf8_1.2.3             lattice_0.20-45        bslib_0.4.2           
## [148] spatstat.sparse_3.0-0  curl_5.0.0             leiden_0.4.3          
## [151] GO.db_3.13.0           openssl_2.0.5          interp_1.1-3          
## [154] survival_3.5-3         limma_3.48.3           rmarkdown_2.20        
## [157] munsell_0.5.0          fastcluster_1.2.3      GenomeInfoDbData_1.2.6
## [160] iterators_1.0.14       impute_1.66.0          reshape2_1.4.4        
## [163] gtable_0.3.1</code></pre>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
